Adopt std::span in CryptographicallyRandomNumber

cdumez opened the pull request:
#### a7472bcca17d5e3401523e9538f31d883e478fed
<pre>
Adopt std::span in CryptographicallyRandomNumber
<a href="https://bugs.webkit.org/show_bug.cgi?id=272635">https://bugs.webkit.org/show_bug.cgi?id=272635</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/CryptographicallyRandomNumber.cpp:
(WTF::cryptographicallyRandomValues):
* Source/WTF/wtf/CryptographicallyRandomNumber.h:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::makeSalt):
* Source/WTF/wtf/OSRandomSource.cpp:
(WTF::cryptographicallyRandomValuesFromOS):
* Source/WTF/wtf/OSRandomSource.h:
* Source/WTF/wtf/RandomDevice.cpp:
(WTF::RandomDevice::cryptographicallyRandomValues):
* Source/WTF/wtf/RandomDevice.h:
* Source/WTF/wtf/UUID.cpp:
(WTF::generateCryptographicallyRandomUUIDVersion4):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::generateTemporaryPath):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::generateCertificate):
* Source/WebCore/Modules/push-api/PushMessageCrypto.cpp:
(WebCore::PushCrypto::ClientKeys::generate):
* Source/WebCore/Modules/websockets/WebSocketFrame.cpp:
(WebCore::appendFramePayload):
* Source/WebCore/Modules/websockets/WebSocketHandshake.cpp:
(WebCore::generateSecWebSocketKey):
* Source/WebCore/crypto/CryptoKey.cpp:
(WebCore::CryptoKey::randomData):
* Source/WebCore/loader/archive/mhtml/MHTMLArchive.cpp:
(WebCore::generateRandomBoundary):
* Source/WebCore/page/Crypto.cpp:
(WebCore::Crypto::getRandomValues):
* Source/WebKit/Shared/IPCTester.cpp:
* Source/WebKit/UIProcess/DeviceIdHashSaltStorage.cpp:
(WebKit::DeviceIdHashSaltStorage::completeDeviceIdHashSaltForOriginCall):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::ephemeralDeviceHashSaltForFrame):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidConnection.cpp:
(WebKit::MockHidConnection::feedReports):
* Source/WebKit/UIProcess/WebAuthentication/Virtual/VirtualHidConnection.cpp:
(WebKit::VirtualHidConnection::parseRequest):

Canonical link: <a href="https://commits.webkit.org/277477@main">https://commits.webkit.org/277477@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b68f91be5b816a4c4221114e18a7e43e9acced4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47732 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26921 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/50495 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/50415 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43787 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/50039 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/32775 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/24399 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38851 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/48314 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/24572 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/41211 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/20148 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/22053 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/42402 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5781 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/41014 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/44076 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/42824 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/52309 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/47223 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22768 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/19095 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/46155 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/24040 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/45189 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24828 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54723 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6748 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23760 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11230 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/OSRandomSource.cpp:
```diff
@@ -32,7 +32,7 @@
 namespace WTF {
 
-void cryptographicallyRandomValuesFromOS(unsigned char* buffer, size_t length)
+void cryptographicallyRandomValuesFromOS(std::span<unsigned char> buffer)
```
`uint8_t`?

darinadler commented on Source/WTF/wtf/OSRandomSource.h:
```diff
@@ -31,5 +31,5 @@ namespace WTF {
 // partially filled.  Rather than calling this function directly, consider
 // calling cryptographicallyRandomNumber or cryptographicallyRandomValues.
-void cryptographicallyRandomValuesFromOS(unsigned char* buffer, size_t length);
+void cryptographicallyRandomValuesFromOS(std::span<unsigned char>);
```
`uint8_t`?

darinadler commented on Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
```diff
@@ -562,8 +562,8 @@ String UserMediaPermissionRequestManagerProxy::ephemeralDeviceHashSaltForFrame(W
-    uint64_t randomData[randomDataSize];
-    cryptographicallyRandomValues(reinterpret_cast<unsigned char*>(randomData), sizeof(randomData));
+    std::array<uint64_t, randomDataSize> randomData;
+    cryptographicallyRandomValues({ reinterpret_cast<uint8_t*>(randomData.data()), sizeof(randomData) });
```
This idiom comes up multiple times. The `std::as_writeable_bytes` function is for cases like this, but it uses `std::byte`, not `uint8_t`. Maybe we should make our own version or even consider using `std::byte` some day.

cdumez replied:
Ok, I'm thinking of adding this to Algorithms.h (until we can move to StdLibExtras.h):
```
template<typename T>
std::span<const uint8_t> asBytes(std::span<T> span)
{
    return { reinterpret_cast<const uint8_t*>(span.data()), span.size() * sizeof(T) };
}

template<typename T>
std::span<uint8_t> asWritableBytes(std::span<T> span)
{
    return { reinterpret_cast<uint8_t*>(span.data()), span.size() * sizeof(T) };
}
```

cdumez committed:
Adopt std::span in CryptographicallyRandomNumber
https://bugs.webkit.org/show_bug.cgi?id=272635

Reviewed by Darin Adler.

* Source/WTF/wtf/CryptographicallyRandomNumber.cpp:
(WTF::cryptographicallyRandomValues):
* Source/WTF/wtf/CryptographicallyRandomNumber.h:
* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::makeSalt):
* Source/WTF/wtf/OSRandomSource.cpp:
(WTF::cryptographicallyRandomValuesFromOS):
* Source/WTF/wtf/OSRandomSource.h:
* Source/WTF/wtf/RandomDevice.cpp:
(WTF::RandomDevice::cryptographicallyRandomValues):
* Source/WTF/wtf/RandomDevice.h:
* Source/WTF/wtf/UUID.cpp:
(WTF::generateCryptographicallyRandomUUIDVersion4):
* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::generateTemporaryPath):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::generateCertificate):
* Source/WebCore/Modules/push-api/PushMessageCrypto.cpp:
(WebCore::PushCrypto::ClientKeys::generate):
* Source/WebCore/Modules/websockets/WebSocketFrame.cpp:
(WebCore::appendFramePayload):
* Source/WebCore/Modules/websockets/WebSocketHandshake.cpp:
(WebCore::generateSecWebSocketKey):
* Source/WebCore/crypto/CryptoKey.cpp:
(WebCore::CryptoKey::randomData):
* Source/WebCore/loader/archive/mhtml/MHTMLArchive.cpp:
(WebCore::generateRandomBoundary):
* Source/WebCore/page/Crypto.cpp:
(WebCore::Crypto::getRandomValues):
* Source/WebKit/Shared/IPCTester.cpp:
* Source/WebKit/UIProcess/DeviceIdHashSaltStorage.cpp:
(WebKit::DeviceIdHashSaltStorage::completeDeviceIdHashSaltForOriginCall):
* Source/WebKit/UIProcess/UserMediaPermissionRequestManagerProxy.cpp:
(WebKit::UserMediaPermissionRequestManagerProxy::ephemeralDeviceHashSaltForFrame):
* Source/WebKit/UIProcess/WebAuthentication/Mock/MockHidConnection.cpp:
(WebKit::MockHidConnection::feedReports):
* Source/WebKit/UIProcess/WebAuthentication/Virtual/VirtualHidConnection.cpp:
(WebKit::VirtualHidConnection::parseRequest):

Canonical link: https://commits.webkit.org/277477@main
